### PR TITLE
Doc for deploying static ray cluster to K8S with external Redis for fault tolerance

### LINKED ISSUE #757: [Feature] [Docs] Docs on how to deploy on K8s without KubeRay
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

Strictly speaking, this is not a KubeRay issue. However, I think it makes sense to track anything related to Ray on Kubernetes in the KubeRay repo.

Many users do not have the cluster-scoped permissions required to install the KubeRay CRDs in their K8s cluster. Other users prefer not to use custom controllers for natural reasons of risk aversion -- Kubernetes built-ins are generally more stable and reliable than community-maintained custom controllers.

For this reason, there should be a page of docs explaining how to manage a Ray cluster by manually creating deployments and services. 
There are actually docs on this in older releases of the Ray repo:
https://github.com/ray-project/ray/blob/releases/1.0.0/doc/source/cluster/kubernetes.rst

Should probably mention the downsides of not using KubeRay:
- No access to bells and whistles (autoscaling, fault-tolerant services, simplified config interfaces, etc.)
- More manual work required.
- KubeRay is the official solution, so other methods will not be as well supported.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by YQ-Wang**
## Why are these changes needed?
This PR adds the documentation and sample config files for deploying Ray to K8S without using KubeRay. As KubeRay CRDs need cluster-scoped permissions, this PR helps those users who do not have cluster-scoped permissions to install Ray Cluster in their K8S.

## Related issue number
Closes [#757](https://github.com/ray-project/kuberay/issues/757)

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [x] This PR is not tested :(

### CODE REVIEW

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (shrekris-anyscale) [inline doc/source/cluster/kubernetes/getting-started.ipynb:320]:
```suggestion
    "Wait for the pods to reach the `Running` state. This may take a few minutes -- most of this time is spent downloading the Ray images. In a separate shell, you may wish to observe the pods' status in real-time with the following command:"
```

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (shrekris-anyscale) [inline doc/source/cluster/kubernetes/getting-started.ipynb:341]:
```suggestion
    "If your pods are stuck in the `Pending` state, you can check for errors via `kubectl describe pod deployment-ray-head-xxxx-xxxxx` and ensure that your Docker resource limits are set high enough."
```

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (shrekris-anyscale) [inline doc/source/cluster/kubernetes/getting-started.ipynb:272]:
```suggestion
    "In this section, we will deploy a static Ray cluster into the `default` namespace without using KubeRay. To use another namespace, specify the namespace in your kubectl commands:\n",
```

Comment (shrekris-anyscale) [reply doc/source/cluster/kubernetes/getting-started.ipynb:272]: Could you also make this change in [the previous subsection](https://github.com/ray-project/ray/pull/31403/files#diff-7480f682fe8e63c4e013a96b1e2e79c7586e6617818eebf5ddd128893498e825R161)?

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (shrekris-anyscale) [inline doc/source/cluster/kubernetes/getting-started.ipynb:509]:
```suggestion
    "First, we need to find the location of the Ray head node. Both the KubeRay operator and the static Ray cluster configuration file configure a [Kubernetes service](https://kubernetes.io/docs/concepts/services-networking/service/) targeting the Ray head pod. This service lets us interact with Ray clusters without directly executing commands in the Ray container. To identify the Ray head service for our example cluster, run:"
```

**Review by shrekris-anyscale — COMMENTED** [review]
Great work on this documentation! I left some comments. Please let me know if anything's unclear.

"Getting Started" documentation link for convenience: https://ray--31403.org.readthedocs.build/en/31403/cluster/kubernetes/getting-started.html

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/getting-started.ipynb:285]: No, we need to first set the real password then run this command.

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/getting-started.ipynb]: I wrote like this to match the existing URLs in the doc. It should be accessible after merging this PR.

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (shrekris-anyscale) [reply doc/source/cluster/kubernetes/getting-started.ipynb:285]: I see– could we change the default `redis-password` in the config to be either empty or `5241590000000000` (Ray's default Redis password)? And then could we add a comment or a note saying the if the command fails, the user should download the config file, change the password, and retry?

Ideally, this command should work out-of-the-box since it'll be confusing if it always fails.

Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/getting-started.ipynb:285]: Done, I set that as an empty string because it is supposed to be the password for the external Redis rather than the default Ray Redis.

File: doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml:4]: Done

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (shrekris-anyscale) [reply doc/source/cluster/kubernetes/getting-started.ipynb:285]: Sounds good!

**Review by shrekris-anyscale — APPROVED** [review]
Thanks for addressing everything in the review!

File: doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml
Comment (simon-mo) [inline doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml:40]: can we simplify the range here? https://kubernetes.io/docs/concepts/services-networking/network-policies/#targeting-a-range-of-ports

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (simon-mo) [inline doc/source/cluster/kubernetes/getting-started.ipynb:416]: you should also discuss the decision of using statically set port ranges for the worker ports. This is done because of the network policy. If there isn't the restriction, the worker port range is not required.

**Review by simon-mo — COMMENTED** [review]
This is awesome and thank you for your contribution!

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/getting-started.ipynb:416]: Good point, let me add the discussion.

File: doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml:40]: I thought about this before. I listed all the ports here mainly for backward compatibility. The range port feature is introduced for the newer version of K8S API, which is not supported in the old versions.

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/getting-started.ipynb:417]: Done, I added the context in [this commit](https://github.com/ray-project/ray/pull/31403/commits/8b9fcfc7dc78a420c0ef718fd1ff2067bad2c086).

**Comment by simon-mo** [comment]
@YQ-Wang, the content itself is in a good state. Thank you for all the prompt edits!

I want to wait for @architkulkarni to finish the final edits. In particular, we need to decide where should this content go. Currently it is under Clusters -> Kubernetes -> KubeRay -> Static Cluster.

The static deployment is not part of KubeRay workflow. We should move the content somewhere else that clearly states:
- This is for cases when KubeRay is not allowed (i.e. no cluster level permission).
- This is advanced content.

Potentially we can put it in a standalone user guide page. But I will leave @architkulkarni to manage this.

**Comment by YQ-Wang** [comment]
> 



> @YQ-Wang, the content itself is in a good state. Thank you for all the prompt edits!
> 
> I want to wait for @architkulkarni to finish the final edits. In particular, we need to decide where should this content go. Currently it is under Clusters -> Kubernetes -> KubeRay -> Static Cluster.
> 
> The static deployment is not part of KubeRay workflow. We should move the content somewhere else that clearly states:
> 
> * This is for cases when KubeRay is not allowed (i.e. no cluster level permission).
> * This is advanced content.
> 
> Potentially we can put it in a standalone user guide page. But I will leave @architkulkarni to manage this.

Thank you for reviewing this PR and giving support!

**Review by architkulkarni — COMMENTED** [review]
This looks great! 

I agree with Simon that this should be a separate page under ["User Guides"](https://ray--31403.org.readthedocs.build/en/31403/cluster/kubernetes/user-guides.html). Perhaps we can title it "(Advanced) Deploying a static Ray cluster without KubeRay".  We can then just link to it from the main KubeRay page where we talk about "two ways to deploy".  

In the new page, we should include a brief high level overview near the top of the page explaining why a user might would want to do this, and what the pros and cons are.  You could probably just do some combination of copying/moving parts of the [paragraph](https://github.com/ray-project/ray/pull/31403/files#diff-7480f682fe8e63c4e013a96b1e2e79c7586e6617818eebf5ddd128893498e825R87) you already wrote, I'll leave it up to you.  Then on the existing page we don't need the new level of subsections and can revert to how it was before.

What do you think?

Apart from that, just minor comments.

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (architkulkarni) [inline doc/source/cluster/kubernetes/getting-started.ipynb:370]: I got a 404, any ideas? https://raw.githubusercontent.com/ray-project/ray/master/doc/source/cluster/kubernetes/configs/static-ray-cluster.with-fault-tolerance.yaml

**Comment by architkulkarni** [comment]
Also, does external Redis only apply to the static approach, or does it apply to both ways?

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/getting-started.ipynb:370]: Same for the other existing doc links, this URL is hard-coded for now. It should be able to work after merging this PR.

Comment (architkulkarni) [reply doc/source/cluster/kubernetes/getting-started.ipynb:370]: Ah, gotcha!

File: doc/source/cluster/kubernetes/getting-started.ipynb
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/getting-started.ipynb]: Done, I added more description and downside effect.

File: doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml
Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/configs/static-ray-cluster-networkpolicy.yaml:39]: Done, I added more contexts [here](https://github.com/ray-project/ray/pull/31403/commits/e12a4780801944cb58c7114182822f43c7fa0a1b#diff-faf3cabc07b24c76a81c1a25fb2b5a5ad1772dd11dfdb511396073c5639d9918R4).

**Comment by YQ-Wang** [comment]
> 



> Also, does external Redis only apply to the static approach, or does it apply to both ways?

@architkulkarni I think the KubeRay approach requires external Redis for fault tolerance by default ([doc](https://ray-project.github.io/kuberay/guidance/gcs-ft/#use-external-redis-cluster)).

**Comment by YQ-Wang** [comment]
> 



> This looks great!
> 
> I agree with Simon that this should be a separate page under ["User Guides"](https://ray--31403.org.readthedocs.build/en/31403/cluster/kubernetes/user-guides.html). Perhaps we can title it "(Advanced) Deploying a static Ray cluster without KubeRay". We can then just link to it from the main KubeRay page where we talk about "two ways to deploy".
> 
> In the new page, we should include a brief high level overview near the top of the page explaining why a user might would want to do this, and what the pros and cons are. You could probably just do some combination of copying/moving parts of the [paragraph](https://github.com/ray-project/ray/pull/31403/files#diff-7480f682fe8e63c4e013a96b1e2e79c7586e6617818eebf5ddd128893498e825R87) you already wrote, I'll leave it up to you. Then on the existing page we don't need the new level of subsections and can revert to how it was before.
> 
> What do you think?
> 
> Apart from that, just minor comments.

@architkulkarni Done, I described the context on a new page. Thanks!

**Comment by architkulkarni** [comment]
@YQ-Wang Thanks! Could you double check the doc build? It looks like a relevant failure: https://buildkite.com/ray-project/oss-ci-build-pr/builds/10179#0185e1c8-ab57-41ee-861a-7388a477cbac/1690-2078

File: doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md
Comment (architkulkarni) [inline doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md:11]: Is this page missing the rest of the page? It only has an overview currently

Comment (YQ-Wang) [reply doc/source/cluster/kubernetes/user-guides/static-ray-cluster-without-kuberay.md:10]: Done, I migrated the contents from the getting-started to here.

**Pull request closed without merging**
